pmlc/dialect/stdx/transforms/subgroup_broadcast.cc

```diff
@@ -0,0 +1,207 @@
+// Copyright 2020 Intel Corporation
+
+#include "mlir/Dialect/GPU/GPUDialect.h"
+#include "mlir/Dialect/GPU/ParallelLoopMapper.h"
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Dialect/Vector/VectorOps.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/StandardTypes.h"
+#include "mlir/Pass/Pass.h"
+
+#include "pmlc/dialect/stdx/ir/ops.h"
+#include "pmlc/dialect/stdx/transforms/pass_detail.h"
+#include "pmlc/util/logging.h"
+
+using namespace mlir; // NOLINT[build/namespaces]
+
+namespace pmlc::dialect::stdx {
+
+namespace {
+
+/// Finds proper sequence of vector operations, performs unvectorization and
+/// adds stdx.subgroup_broadcast
+///
+/// Sample code:
+///
+///  %c1_i32 = constant 1 : i32
+///  %c0 = constant 0 : index
+///  %c8 = constant 8 : index
+///  %c64 = constant 64 : index
+///  %cst = constant 0.000000e+00 : f32
+///  scf.parallel (%i) = (%c0) to (%c64) step (%c8) {
+///    %0 = vector.transfer_read %arg0[%i], %cst : memref<64xf32>, vector<8xf32>
+///    %1 = vector.extractelement %0[%c1_i32 : i32] : vector<8xf32>
+///    %2 = vector.broadcast %1 : f32 to vector<8xf32>
+///    vector.transfer_write %2, %arg1[%i] : vector<8xf32>, memref<64xf32>
+///    scf.yield
+///  }
+///
+/// Code after transformation:
+///
+///  %c1_i32 = constant 1 : i32
+///  %c0 = constant 0 : index
+///  %c1 = constant 1 : index
+///  %c8 = constant 8 : index
+///  %c64 = constant 64 : index
+///  scf.parallel (%i, %sid) = (%c0, %c0) to (%c64, %c8) step (%c8, %c1) {
+///    %idx = addi %i, %sid : index
+///    %0 = load %arg0[%idx] : memref<64xf32>
+///    %1 = stdx.subgroup_broadcast %0, %c1_i32 : f32
+///    store %1, %arg1[%idx] : memref<64xf32>
+///    scf.yield
+///  } { mapping = [
+///    {processor = 1, map = affine_map<(d0) -> (d0)>, bound = affine_map<(d0)
+///    -> (d0)>}, {processor = 0, map = affine_map<(d0) -> (d0)>, bound =
+///    affine_map<(d0) -> (d0)>}] }
+///
+/// This pass requires --convert-parallel-loops-to-gpu afterwards
+
+void runOnParallel(scf::ParallelOp loopOp) {
```

Comment:
The pattern matching here is too specific.  Additionally, the 'devectorization' needs to happen universally thoughout the scf.parallel (i.e. all vector reads/writes need to be 'scalarized', and vector computational operations (i.e. mulf) need to be retyped to operate on scalars.  Specifically the transforms need to be roughly:

1) transfer_read X[i, j, k] -> load[i, j, k + sid]
2) transfer_write  a, X[i, j, k] -> store a, X[i, j, k + sid]
3) b = extractelement a[e] vec, broadcast b -> stdx.subgroup_broadcast a, e
4) Change the typing of any vector versions of std.mulf, etc, with scalar versions of the same.  You may want to rely on the op implementing VectorUnrollOpInterface to detect such ops as is done here: https://github.com/plaidml/plaidml/blob/plaidml-v1/pmlc/dialect/pxa/transforms/vectorize.cc#L84

Please let me know if this makes sense.  I can try to provide a more complete set of lit test by EOD Tuesday if that would help.  There is an open question in my mind regarding two points:

1) Should we consider converting adding some sort of attribute to the the load/store resulting from  the transfer_read/transfer_write so that it is marked for lowering to sub_group_block_read/sub_group_block_write or should we rely on the later lowering to detect this due to the stride 1 relative to SID + the alignment?

2) Is it reasonable long term 'devectorize' all computational ops greedily?  This means that any use of vectorization of ops on a per-thread basis would need to done after the subgroup lowering.  To reproduce the behavior of plaidml-v0, which had excellent performance, we can do things greedily, but I wonder if we are painting ourselves into a corner.